core/src/main/java/com/linecorp/armeria/server/HttpHeaderUtil.java

```diff
@@ -174,4 +177,58 @@ static void ensureUniqueMediaTypes(Iterable<MediaType> types, String typeName) {
     }
 
     private HttpHeaderUtil() {}
+
+    static String maybeTransformAbsoluteUri(
+            String path, Function<? super String, String> absoluteUriTransformer) throws URISyntaxException {
+
+        if (isValidHttp2Path(path)) {
+            return path;
+        }
+
+        if (!ArmeriaHttpUtil.isAbsoluteUri(path)) {
+            throw newInvalidPathException(path);
+        }
+
+        final String newPath;
+        try {
+            newPath = absoluteUriTransformer.apply(path);
+        } catch (Exception e) {
+            warnExceptionThrowingAbsoluteUriTransformer(e);
+            throw newInvalidPathException(path);
+        }
+
+        if (newPath == null || newPath.isEmpty()) {
+            warnNullReturningAbsoluteUriTransformer();
+            throw newInvalidPathException(path);
+        }
+
+        if (!isValidHttp2Path(newPath)) {
+            throw newInvalidPathException(path);
+        }
+
+        return newPath;
+    }
+
+    private static void warnExceptionThrowingAbsoluteUriTransformer(Exception e) {
+        if (!warnedIllegalAbsoluteUriTransformer) {
+            warnedIllegalAbsoluteUriTransformer = true;
+            logger.warn("absoluteUriTransformer.apply() raised an exception; returning 400 Bad Request", e);
+        }
+    }
+
+    private static void warnNullReturningAbsoluteUriTransformer() {
+        if (!warnedIllegalAbsoluteUriTransformer) {
+            warnedIllegalAbsoluteUriTransformer = true;
+            logger.warn("absoluteUriTransformer.apply() returned null; returning 400 Bad Request");
+        }
+    }
+
+    private static URISyntaxException newInvalidPathException(String path) {
+        return new URISyntaxException(path, "neither origin form nor asterisk form");
+    }
+
+    private static boolean isValidHttp2Path(String path) {
```

Comment:
This isn't your fault but I think we need to check if the method is OPTION when the path is `*`.
https://www.rfc-editor.org/rfc/rfc7540#section-8.1.2.3

Reply:
I understood your suggestion to mean throw an exception when a path is invalid.
I don't see a reason to only check * with OPTIONS, but we could also short-circuit when `pathAndQuery == null`.

Unfortunately, I think this potentially changes behavior and may print incorrect warning logs.
https://github.com/line/armeria/pull/4721#issuecomment-1481199075

While I think it is a good idea to throw an exception on path validation, I think the scope of change is large and deserves its own PR.

Reply:
> I don't see a reason to only check * with OPTIONS,

We are doing it here so I thought we needed to apply the same approach:
https://github.com/line/armeria/blob/3f83c66a1c73dea6aac31a6b8072edefbea10e62/core/src/main/java/com/linecorp/armeria/server/ServiceRouteUtil.java#L42

> While I think it is a good idea to throw an exception on path validation, I think the scope of change is large and deserves its own PR.

Yeah, let's solve it in another PR. 👍